modules/v2x/v2x_proxy/app/v2x_proxy.cc

```diff
@@ -13,147 +13,199 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  *****************************************************************************/
-
 /**
  * @file v2x_proxy.cc
  * @brief define v2x proxy class
  */
-
 #include "modules/v2x/v2x_proxy/app/v2x_proxy.h"
 
+#include <cmath>
+#include <cstdio>
+#include <ostream>
+#include <vector>
+
+#include "modules/v2x/common/v2x_proxy_gflags.h"
+
 namespace apollo {
 namespace v2x {
+using ::apollo::canbus::ChassisDetail;
+using ::apollo::v2x::CarStatus;
+using ::apollo::v2x::StatusResponse;
+using ::apollo::v2x::obu::OBURsi;
+using ::apollo::v2x::obu::OBUTrafficLightData;
+
+V2xProxy::~V2xProxy() {
+  if (recv_thread_ != nullptr && recv_thread_->joinable()) {
+    recv_thread_->join();
+  }
+  if (planning_thread_ != nullptr && planning_thread_->joinable()) {
+    planning_thread_->join();
+  }
+  // rsi_thread_->join();
+}
 
-using apollo::localization::LocalizationEstimate;
-
-V2xProxy::V2xProxy()
-    : node_(cyber::CreateNode("v2x_proxy")), init_flag_(false) {
+V2xProxy::V2xProxy(std::shared_ptr<::apollo::hdmap::HDMap> hdmap)
+    : node_(::apollo::cyber::CreateNode("v2x_proxy")), exit_(false) {
   if (node_ == nullptr) {
     AFATAL << "Create v2x proxy node failed.";
     exit(1);
   }
-
-  auto x2v_trafficlight_timer_period = ceil(
-      (1.0 / static_cast<int>(FLAGS_x2v_trafficlight_timer_frequency)) * 1000);
-  x2v_trafficlight_timer_.reset(
-      new cyber::Timer(static_cast<uint32_t>(x2v_trafficlight_timer_period),
-                       [this]() { this->OnX2vTrafficLightTimer(); }, false));
-  auto v2x_carstatus_timer_period = ceil(
-      (1.0 / static_cast<int>(FLAGS_v2x_carstatus_timer_frequency)) * 1000);
-  v2x_carstatus_timer_.reset(
-      new cyber::Timer(static_cast<uint32_t>(v2x_carstatus_timer_period),
-                       [this]() { this->OnV2xCarStatusTimer(); }, false));
-
+  internal_ = std::make_shared<InternalData>();
+  hdmap_ = std::make_shared<::apollo::hdmap::HDMap>();
+  const auto hdmap_file = apollo::hdmap::BaseMapFile();
+  if (0 != hdmap_->LoadMapFromFile(hdmap_file)) {
+    AERROR << "Failed to load hadmap file: " << hdmap_file;
+    return;
+  }
+  auto v2x_car_status_timer_period = std::ceil(
+      (1.0 / static_cast<int>(FLAGS_v2x_car_status_timer_frequency)) * 1000);
+  ::apollo::cyber::TimerOption v2x_car_status_timer_option;
+  v2x_car_status_timer_option.period =
+      static_cast<uint32_t>(v2x_car_status_timer_period);
+  v2x_car_status_timer_option.callback = [this]() {
```

Comment:
Why static_cast forth and back? `auto period = static_cast<uint32_t>((1000 + freq - 1 ) / freq); ` is enough to produce the same result.